clang/lib/CIR/CodeGen/CIRGenBuilder.cpp

```diff
@@ -0,0 +1,66 @@
+//===-- CIRGenBuilder.cpp - CIRBuilder implementation ---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include "CIRGenBuilder.h"
+
+namespace cir {
+
+mlir::Value CIRGenBuilderTy::maybeBuildArrayDecay(mlir::Location loc,
+                                                  mlir::Value arrayPtr,
+                                                  mlir::Type eltTy) {
+  auto arrayPtrTy =
+      ::mlir::dyn_cast<::mlir::cir::PointerType>(arrayPtr.getType());
+  assert(arrayPtrTy && "expected pointer type");
+  auto arrayTy =
+      ::mlir::dyn_cast<::mlir::cir::ArrayType>(arrayPtrTy.getPointee());
+
+  if (arrayTy) {
+    mlir::cir::PointerType flatPtrTy =
+        mlir::cir::PointerType::get(getContext(), arrayTy.getEltType());
+    return create<mlir::cir::CastOp>(
+        loc, flatPtrTy, mlir::cir::CastKind::array_to_ptrdecay, arrayPtr);
+  }
+
+  assert(arrayPtrTy.getPointee() == eltTy &&
+         "flat pointee type must match original array element type");
+  return arrayPtr;
+}
+
+mlir::Value CIRGenBuilderTy::buildArrayAccessOp(
+    mlir::Location arrayLocBegin, mlir::Location arrayLocEnd,
+    mlir::Value arrayPtr, mlir::Type eltTy, mlir::Value idx, bool shouldDecay) {
+  mlir::Value basePtr = arrayPtr;
+  if (shouldDecay)
+    basePtr = maybeBuildArrayDecay(arrayLocBegin, arrayPtr, eltTy);
```

Comment:
- We don't need two functions, please change the faster path to `if (!shouldDecay)` and inline the other method after it. 
- Rename `buildArrayAccessOp` to `getArrayElement`. If `arrayPtr` is always a pointer, it should be `mlir::cir::PointerType`.

Reply:
The function maybeBuildArrayDecay is called elsewhere 
https://github.com/llvm/clangir/blob/828b08997ea4b03a524554bc1c88a7679974f663/clang/lib/CIR/CodeGen/CIRGenExpr.cpp#L1569
So it cannot be simply inlined here unless we are willing to inline at the other call site. 

arrayPtr is of type mlir::Value, what does it mean to make it mlir::cir::PointerType?

Reply:
Looks like I missed the other callsite & read mlir::Type instead of mlir::Value, nothing to do here then.